File: src/main/lua/SimpleL2/CHIBridge.lua
local utils = require "LuaUtils"
local env = require "env"

local assert = assert
local expect = env.expect

local bridge = dut.u_CHIBridge
local txState = bridge.txState
local rxState = bridge.rxState
local txactivereq = dut.io_out_chiLinkCtrl_txactivereq
local txactiveack = dut.io_out_chiLinkCtrl_txactiveack
local rxactivereq = dut.io_out_chiLinkCtrl_rxactivereq
local rxactiveack = dut.io_out_chiLinkCtrl_rxactiveack

assert(cfg.simulator == "vcs" or cfg.simulator == "iverilog")

local LinkState = utils.enum_define {
    STOP = 0,
    ACITVATE = 2,
    RUN = 3,
    DEACTIVATE = 1,
}

local test_txstate_switch = env.register_test_case "test_txstate_switch" {
    function ()
        env.dut_reset()

        env.posedge(3)
            txState:expect(LinkState.ACITVATE)
        
        env.posedge()
            txState:expect(LinkState.ACITVATE)
            txactiveack:set(1)
        
        env.posedge(2)
            txState:expect(LinkState.RUN)
            txactivereq:expect(1)
        
        env.posedge()

        env.posedge(10, function (cycle)
            txState:expect(LinkState.RUN) -- keep RUN
            txactivereq:expect(1)
            txactiveack:expect(1)
        end)

        -- txactivereq:set_force(0)
        -- txactiveack:set_force(0)

        -- TODO: STOP
    end
}

local test_txstate_after_reset = env.register_test_case "test_txstate_after_reset" {
    function ()
        env.dut_reset()
        txactiveack:set(0)
        env.posedge(3)

        txState:expect(LinkState.ACITVATE)

        env.posedge(10, function (cycle)
            txState:expect(LinkState.ACITVATE) -- keep ACITVATE
        end)
    end
}

local test_rxstate_switch = env.register_test_case "test_rxstate_switch" {
    function ()
        env.dut_reset()
            rxState:expect(LinkState.STOP)
        
        env.posedge(3)
            rxState:expect(LinkState.STOP)
            rxactiveack:expect(0)
            rxactivereq:expect(0)

            rxactivereq:set(1)
        
        env.posedge(2)
            rxState:expect(LinkState.ACITVATE)
        
        env.posedge()
            rxState:expect(LinkState.RUN)

        env.posedge(10, function (cycle)
            assert(rxState() == LinkState.RUN) -- keep RUN
            assert(rxactiveack() == 1)
            assert(rxactivereq() == 1)
        end)

        rxactivereq:set(0)

        env.posedge(2)
            rxState:expect(LinkState.DEACTIVATE)
        
        env.posedge()
            rxState:expect(LinkState.STOP)

        env.posedge(10, function (cycle)
            rxState:expect(LinkState.STOP) -- keep STOP
            rxactiveack:expect(0)
            rxactivereq:expect(0)
        end)

        
        -- 
        -- restart
        -- 
        rxactivereq:set(1)
        env.posedge(2)
            rxState:expect(LinkState.ACITVATE)

        env.posedge()
            rxState:expect(LinkState.RUN)

        env.posedge(10, function (cycle)
            rxState:expect(LinkState.RUN) -- keep RUN
            rxactiveack:expect(1)
            rxactivereq:expect(1)
        end)

        rxactivereq:set(0)

        env.posedge(10)
            rxState:expect(LinkState.STOP)
    end
}

verilua "mainTask" { function ()
    sim.dump_wave()
    env.dut_reset()

    test_txstate_switch()
    test_txstate_after_reset()
    test_rxstate_switch()

    env.TEST_SUCCESS()
end }

File: src/main/scala/SimpleL2/CHIBridge.scala
package SimpleL2

import chisel3._
import chisel3.util._
import org.chipsalliance.cde.config._
import xs.utils.perf.{DebugOptions, DebugOptionsKey}
import Utils.GenerateVerilog
import SimpleL2.Configs._
import SimpleL2.chi._

object LinkState {
    def STOP = "b00".U(2.W)
    def ACTIVATE = "b10".U(2.W)
    def RUN = "b11".U(2.W)
    def DEACTIVATE = "b01".U(2.W)
}

class CHIBridgeInput(implicit p: Parameters) extends L2Bundle {
    val chi = CHIBundleDecoupled(chiBundleParams)
}

class CHIBridgeOutput(implicit p: Parameters) extends L2Bundle {
    val chi         = CHIBundleDownstream(chiBundleParams)
    val chiLinkCtrl = new CHILinkCtrlIO()
}

class CHIBridge()(implicit p: Parameters) extends L2Module {
    val io = IO(new Bundle {
        // val in  = new CHIBridgeInput
        val out = new CHIBridgeOutput
        // TODO: val deactivateTxLink = Input(Bool())
        // TODO: ShutDown
    })

    io.out.chi         <> DontCare
    io.out.chiLinkCtrl <> DontCare

    val txactivereq = io.out.chiLinkCtrl.txactivereq
    val txactiveack = io.out.chiLinkCtrl.txactiveack
    val rxactivereq = io.out.chiLinkCtrl.rxactivereq
    val rxactiveack = io.out.chiLinkCtrl.rxactiveack

    val resetFinish = RegInit(false.B)
    resetFinish := true.B

    val txState = RegInit(LinkState.STOP)
    val rxState = RegInit(LinkState.STOP)
    dontTouch(txState)
    dontTouch(rxState)

    val updateTxState = WireInit(false.B)
    val updateRxState = WireInit(false.B)
    dontTouch(updateTxState)
    dontTouch(updateRxState)

    // @formatter:off
    updateTxState := (RegNext(RegNext(reset.asBool)) && resetFinish)                    ||
                     (txactivereq  && txactiveack && txState === LinkState.ACTIVATE)    ||
                     (!txactivereq && txactiveack && txState === LinkState.RUN)         ||
                     (!txactivereq && !txactiveack && txState === LinkState.DEACTIVATE)
                    //  TODO: LinkState.STOP
    // @formatter:on

    when(updateTxState && resetFinish) {
        when(txState === LinkState.STOP) {
            txState := LinkState.ACTIVATE
        }.elsewhen(txState === LinkState.ACTIVATE) {
            txState := LinkState.RUN
        }.elsewhen(txState === LinkState.RUN) {
            txState := LinkState.DEACTIVATE
        }.elsewhen(txState === LinkState.DEACTIVATE) {
            txState := LinkState.STOP
        }
    }

    // @formatter:off
    updateRxState := !reset.asBool && (
                        (rxactivereq && rxState === LinkState.STOP)        ||
                        (rxactivereq && rxactiveack && rxState === LinkState.ACTIVATE)     ||
                        (!rxactivereq && rxactiveack && rxState === LinkState.RUN)         ||
                        (!rxactivereq && !rxactiveack && rxState === LinkState.DEACTIVATE)
                    )
    // @formatter:on

    when(updateRxState) {
        when(rxState === LinkState.STOP) {
            rxState := LinkState.ACTIVATE
        }.elsewhen(rxState === LinkState.ACTIVATE) {
            rxState := LinkState.RUN
        }.elsewhen(rxState === LinkState.RUN) {
            rxState := LinkState.DEACTIVATE
        }.elsewhen(rxState === LinkState.DEACTIVATE) {
            rxState := LinkState.STOP
        }
    }

    io.out.chiLinkCtrl.txactivereq := txState === LinkState.ACTIVATE || txState === LinkState.RUN
    io.out.chiLinkCtrl.rxactiveack := rxactivereq && rxState === LinkState.STOP || rxState === LinkState.ACTIVATE || rxState === LinkState.RUN

    assert(!(txState === LinkState.RUN && txactivereq && !txactiveack), "txactiveack should keep high during RUN state")
    assert(!(txState === LinkState.ACTIVATE && !txactivereq), "txactivereq should keep high during ACTIVATE state")

    dontTouch(io)
}

object CHIBridge extends App {
    val config = new Config((_, _, _) => {
        case L2ParamKey      => L2Param()
        case DebugOptionsKey => DebugOptions()
    })

    GenerateVerilog(args, () => new CHIBridge()(config), name = "CHIBridge", split = false)
}
